daemons/ipa-otpd/passkey.c

```diff
@@ -0,0 +1,823 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This file contains various helper functions for the passkey feature.
+ */
+
+#define _GNU_SOURCE /* for asprintf() */
+#include <stdio.h>
+#include <fcntl.h>
+#include <jansson.h>
+#include <openssl/rand.h>
+#include <openssl/evp.h>
+
+#include "internal.h"
+
+struct passkey_data {
+    int phase;
+    char *state;
+    union {
+        struct passkey_challenge {
+            char *domain;
+            json_t *credential_id_list;
+            int user_verification;
+            unsigned char *cryptographic_challenge;
+        } challenge;
+
+        struct sss_passkey_reply {
+            char *credential_id;
+            char *cryptographic_challenge;
+            char *authenticator_data;
+            char *assertion_signature;
+        } response;
+    } data;
+    json_t *jdata;
+    json_t *jroot;
+};
+
+struct otpd_queue_item_passkey {
+    char *domain;
+    char *ipaRequireUserVerification;
+    struct passkey_data *data_in;
+    struct passkey_data *data_out;
+    krb5_data state;
+};
+
+static void free_passkey_data(struct passkey_data *p)
+{
+    if (p == NULL) {
+        return;
+    }
+
+    if (p->phase == 1) {
+        free(p->data.challenge.domain);
+        free(p->data.challenge.cryptographic_challenge);
+    }
+
+    json_decref(p->jdata);
+    json_decref(p->jroot);
+    free(p);
+}
+
+void free_otpd_queue_item_passkey(struct otpd_queue_item *item)
+{
+    free(item->passkey->domain);
+    free(item->passkey->ipaRequireUserVerification);
+
+    free_passkey_data(item->passkey->data_in);
+    free_passkey_data(item->passkey->data_out);
+
+    free(item->passkey);
+}
+
+static struct otpd_queue_item_passkey *get_otpd_queue_item_passkey(void)
+{
+    struct otpd_queue_item_passkey *p;
+
+    p = calloc(1, sizeof(struct otpd_queue_item_passkey));
+    if (p == NULL) {
+        return NULL;
+    }
+
+    p->data_in = calloc(1, sizeof(struct passkey_data));
+    if (p->data_in == NULL) {
+        free(p);
+        return NULL;
+    }
+
+    p->data_out = calloc(1, sizeof(struct passkey_data));
+    if (p->data_out == NULL) {
+        free(p->data_in);
+        free(p);
+        return NULL;
+    }
+
+    p->data_in->phase = -1;
+    p->data_out->phase = -1;
+
+    return p;
+}
+
+#define PASSKEY_PREFIX "passkey "
+
+static const krb5_data *get_full_krad_attr(krad_packet *req, krad_attr attr)
```

Comment:
May be it makes also sense to extract `get_full_krad_attr()` from here and unify with `get_krad_attr_from_packet()` in a common place like you did with `add_krad_attr_to_set()`?

Reply:
fixed